### ISSUE TITLE
Proposal for revamped docs pipeline, with example repo

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#3930]
## 📚 Documentation

I'd like to make a concrete proposal for how our docs are organized on github, built, and published.  I know this is a heavily discussed topic, and it's definitely with great appreciation for the previous work done here that I propose this.  I also recognize that there is a docs working group that I haven't been able to attend, so please accept my apologies if any toes feel stepped on.  Mostly, I'd like to propose a docs build pipeline/architecture, knowing well that this is complicated, and welcoming counter-opinions on anything I've overlooked here.

## need for a sphinx theme

For me, one of the things that causes the biggest amount of difficulties is that we no longer have a pip installable sphinx them.  Our theme **consumes the output** of sphinx, rather than being an input to sphinx.  This is what makes it such that building and developing docs locally rarely results in a successful docs publish after PR.  Nearly every PR for the last year or so has led to a followup "fix docs" PR.

In discussion about how to let people see what the final docs will look like, things like netlify have come up.  To me, that just adds more complexity on top of the existing complexity.  If our docs had complex needs, it might be warranted, but they don't: they can very easily be purely static (and probably *should* be unless they really need otherwise).

So the **primary** requirement here is that our theme be a  **pip installable sphinx theme**, such that someone building locally can just `make docs`, see the result, and expect it to be the same as what happens when github actions does the same.
This also means that the result should effectively be a static site (no AJAX, no react, etc...).  We currently use react, but don't particularly need the power of react (we don't have any interactivity, and don't really have much need for it.  Versioned docs and simple javascript interactivity can definitely be achieved with static pages, so the two should not be conflated).

## steps required to use our current theme

From conversations, I think the biggest disappointment here would probably be that our docs wouldn't match the look and feel of the hub, (until a sphinx theme was created).  Personally, I don't feel like the usage documentation _needs_ to match the hub theme (and I do kind of like matching the general look & feel of other scientific python packages), but I can certainly understand if others feel differently on that, and I know a lot of work has gone into the current theme.  (Again, it's difficult to even suggest this...)

I recognize that's a big lift for our current theme, but when I weigh the aesthetic benefits of our current theme against the practical and structural difficulties we've had with docs this year, it feels like the right call (if perhaps just temporarily).  We could, for now, return to `furo` or `pydata-sphinx-theme`, etc... And if there is a desire to recover the look and feel of our current site, then it could be built as a sphinx theme (i.e. providing templates, CSS, and javascript).  From a webdev perspective, I know that that's a wholly unsatisfying set of tooling, and no modern interactive site would build on that sort of tech.  But a documentation site has different needs, and the complexity of the current pipeline is proving challenging for napari's needs.

## proposal & example repo

I made an example git repo that has :  
https://github.com/napari-docs-example/napari-docs-example.github.io

- This is a *single* place for documentation (with the exception of API docs, which come from docstrings).  This would replace `napari.github.io` (which could then turn into a repository developing a sphinx theme that we could eventually use), and it would remove the `docs` folder and associated tooling from napari.
- What you see when you clone, modify, and run `make docs` is exactly what you'll see when your PR gets merged.  
- It includes all of the tutorials, API docs, and everything.  
- It includes images without the need to upload them elsewhere and hotlink them.
- It automatically builds upon merge to main and publishes to the `gh-pages` branch..

The resulting published site is here: 
https://napari-docs-example.github.io/

with the exception of API docs, it's also building without any warnings (which @melissawm would probably appreciate, and would help us to catch errors).

## one repo without API docs vs docs in napari

@jni had expressed a desire to _not_ create a new repo for docs, which I do get.  But that leaves us with the repo size & images problem. Though it would be an awesome goal, it's not likely that we're going to be able to create a framework that lets us autogenerate _everything_ we'd like to be able to show on our website (at least, not anytime soon), such as tutorial-oriented movies.  So if we stick with `docs` in napari, we're stuck finding some external place to host images, which has its own set of difficulties.  Even if we created a separate repo just for media, every docs PR with media would essentially be two PRs to two repos, trying to coordinate between them.  Building locally gets more complicated, etc...  it's just not awesome.

If we _really_ wanted to maintain the feeling of a docs folder in the napari repo, we could include the "napari-docs" repo as a git submodule in the napari repo, then someone could just fetch submodules and build.

The main downside here, as I see it, is that API docs PRs go to napari, and everything else goes to napari-docs.  But I think, many people are used to the concept of API docs coming from source.  And if not, it's not a hard thing to explain to a newcomer.

## what this isn't

this _isn't_ a suggestion on the structural organization of our docs (diataxis, etc...), just a request/proposal on the build/publish pipeline and repo organization.

## Thoughts?

Anyway, I hope you'll have a look at the [example repo](https://github.com/napari-docs-example/napari-docs-example.github.io), perhaps clone and run `make docs`, or look at the [gh-pages build](https://napari-docs-example.github.io/), and list your concerns or thoughts here.  thanks

cc @jni, @sofroniewn, @melissawm, @neuromusic, @GenevieveBuckley, @DragaDoncila

**Comment by sofroniewn** [comment]
Thanks for writing up the proposal.

> Nearly every PR for the last year or so has led to a followup "fix docs" PR.

Unfortunately this has been my experience too (for PRs with large/ structural docs changes) (see #3927 as a recent example).

It seems like this would be something good to discuss at the next docs working group meeting. It looks like the next one is [Wed 19th](https://napari.org/community/meeting_schedule.html). I can try and attend then too

**Comment by codemonkey800** [comment]
I'm glad I made it easy to pull out of the current theme lol 

I think a lot of this makes sense. A big engineering decision for using React was to support interactive features we'd add in the future and enable code sharing with napari hub because of how similar the theme is to the hub. Versioned docs, the calendar, download matrix, and I think gifs of examples were things I had in mind. The developer tooling ([fast refresh](https://nextjs.org/docs/basic-features/fast-refresh) for example) and code sharing also allowed us to finish the theme in a short amount of time since we were also still developing features for the napari hub.

I'm fine with this, but it means we'll need to drop the current theme and rewrite it from scratch using HTML, CSS, and JS. We may need to drop some things like the Calendar, unless we look into a hybrid approach of using React just for specific features (which is possible, but I guess at this point I want to know if the community even wants React in the docs).

Thank you for the proposal!

**Comment by codemonkey800** [comment]
> So the primary requirement here is that our theme be a pip installable sphinx theme, such that someone building locally can just make docs

Isn't this something we could do with `git submodules`?

**Comment by tlambert03** [comment]
> > So the primary requirement here is that our theme be a pip installable sphinx theme, such that someone building locally can just make docs
> 
> 
> 
> Isn't this something we could do with `git submodules`?

Conceivably sure, but I haven't seen that pattern for a sphinx theme.  Is there an advantage there you're thinking about?

**Comment by tlambert03** [comment]
> Versioned docs, the calendar, download matrix, and I think gifs of examples were things I had in mind

all good stuff, but versioned docs, the download matrix, and gifs of examples are all definitely possible without react (though I'm sure I'm saying something you already know).  don't know what tooling the calendar uses, so can't comment there, but surely it's possible to embed a calendar without react.  pydata-sphinx-theme does versioning with a little bit of vanilla javascript

**Comment by codemonkey800** [comment]
It's mostly to get around the fact that it's not a pip installable sphinx theme. You can write the makefile to clone the theme repo and build the docs (which I thought @melissawm had in mind)

**Comment by tlambert03** [comment]
i see, so it's just to avoid having to put it on PyPI?  meh :)

(so much python tooling is based on a list of pypi installables... so having that one thing in the pipeline that doesn't fit in with the ecosystem is always a bit annoying.  doable, but only with a good reason why)

**Comment by jni** [comment]
Hi all!

I agree with the general gist of @tlambert03's issue, and I agree with him that a standard sphinx docs repo is preferable to the current situation. I'm a big proponent of "when in Rome...", and there should be a high barrier for doing non-standard stuff. So, I'd be happy enough with that approach.

As he's noted, though, I do prefer having docs together with the code. Part of my preference I'm sure is just habit and ecosystem practice, but I think I can at least rationalise my preference. 😜 

- if you have two repos, the two-PR dance is inevitable: updating an API in a PR to main will require an update to the docs in the docs repo. Ditto for updates that touch the UI.
- increasing separation between docs and code increases the likelihood that docs will fall out of date with the code.
- the best way for docs to be updated together with code is for them to be autogenerated from the code. This is easier to do in a monorepo.

In short, my vision for developing the perfect docs infrastructure is:
- figure out example data storage (CZI provides unlimited storage in Dryad for grantees, for example)
- figure out caching of said data on GitHub Actions
- figure out autogeneration of screenshots (more or less trivial) and movies (less trivial, but progress was discussed [a year ago already](https://github.com/napari/napari/issues/1387#issuecomment-769984161))
- profit 😉 

Once you have all that, it becomes ~easy to generate an entire documentation site that is up-to-date with every PR, and I think that's pretty golden. Yes, there's still the issue of e.g. diagrams, but I honestly don't have a huge concern with storing images in github PR comments — even if the system breaks eventually, the images are then easy to find based on git blame giving you the PR number.

**Comment by kne42** [comment]
I'm not a big fan of this idea. A full sphinx approach would be a major regression from a developmental standpoint. I would sooner have us write our own documentation engine than return to sphinx because it's just that outdated.

I believe that returning to pure sphinx would not be in our best interests long-term because:
1. sphinx has no multi-versioning capabilities. it's either been done via plugins (which, when last I checked, I couldn't find any that worked with the latest version), or by custom solutions such as ReadTheDocs.
2. themes would have to be written in plain HTML/CSS/JS which is less efficient and harder to maintain for web developers. furthermore, anyone without prior experience wishing to contribute to the theme would not be receiving exposure to the latest web technologies, so it's not all that useful to them.

> Unfortunately this has been my experience too (for PRs with large/ structural docs changes) (see #3927 as a recent example).

I would like to point out that this is an issue with not having a unified docs repo (due to the copy docs script messing up) and _not_ anything to do with the theme being built in React.

This brings me on to my next point, which is that I think many of the issues that @tlambert03 has brought up are more so because the docs are not in one unified place and less so because we customized our build process. If everything just lived in a single repo we could just use our custom build and everything would look 100% exactly the same when you build it locally vs. via CI.

**Comment by codemonkey800** [comment]
> so it's just to avoid having to put it on PyPI?

Not necessarily, it's something we had in mind but we need to figure out if it's possible to include node dependencies in the build pipeline. Or  we could download node and its dependencies at runtime, and run the theme build after `jupyterbook build` completes. For example, Furo [overwrites the CSS](https://github.com/pradyunsg/furo/blob/d6b217f8bb021f137928aa30f1b41e829e87d73f/src/furo/__init__.py#L370) for pygments when the build completes. Although that would be an additional layer of complexity.

> (so much python tooling is based on a list of pypi installables... so having that one thing in the pipeline that doesn't fit in with the ecosystem is always a bit annoying. doable, but only with a good reason why)

The main reason for us was to improve developer experience of the theme itself. [Static exports](https://nextjs.org/docs/advanced-features/static-html-export) allows us to produce static HTML, CSS, and JS from React components, and fast refresh makes iteration really fast (No more 5 minute wait times for `jupyterbook build` to complete each time I hit `Cmd + S`) It also made sense from the perspective of the hub team since it means we could work in similar codebases and share code.

**Comment by tlambert03** [comment]
> The main reason for us was to improve developer experience of the theme itself.

that makes lots of sense.  but it does seem like something you could sandbox a bit to your development process.  That is, we don't necessarily need every napari docs contributor to have to deal with the whole node toolset just because the theme developers work in node.

I've built websites elsewhere using react.  as I said earlier, I'm fully aware of how woefully antiquated the conventional sphinx pipeline appears to a modern web dev.  i just think we need to consider the target audience, and the specifics of our site interactivity use case here when considering whether it's worth it (even if it's clearly worth it in other webdev scenarios).

**Comment by codemonkey800** [comment]
If we could package up the theme into a pip installable package that works (regardless of how it gets the node dependencies), would that be a good enough sandbox? Or is there more interest in being able to contribute to the theme without getting into Node + React?

**Comment by tlambert03** [comment]
no no, whatever you want to do on the theme dev side is fine.  you can use whatever tool chain you want there (there's no expectation that anyone has to be able to contribute to the _theme_ easily) ... on our side, the requirements would just be that it is installable as a sphinx theme that we can add to our `_conf.yml`.

as mentioned in the original post, (and this is likely more limiting to you) I would also much prefer that the theme not require someone to spin up a local webserver.  you should just be able to open `file://...` after build

**Comment by codemonkey800** [comment]
> the requirements would just be that it is installable as a sphinx theme that we can add to our _conf.yml.

That's what I meant. We could set up the theme into a pip package that you can just `pip install` and specify in the configuration. You don't need to worry about node or anything, you can just run `jupyter-book build`.  The HTML will then be accessible to you from`file://` (which is possible with the current theme if you do `yarn build`).

**Comment by DragaDoncila** [comment]
> Nearly every PR for the last year or so has led to a followup "fix docs" PR.

I completely agree that the current process for building our docs is untenable, and it's a matter of very high priority to address this issue. I also agree that introducing more tooling e.g. netlify is complicating the matter, and not addressing the fact that PR previews should be in addition to an easy local process, not a replacement for it.

> This also means that the result should effectively be a static site (no AJAX, no react, etc...). We currently use react, but don't particularly need the power of react (we don't have any interactivity, and don't really have much need for it. Versioned docs and simple javascript interactivity can definitely be achieved with static pages, so the two should not be conflated).

I tend to agree with this as well. The power of react comes into play when building websites with many, large interactive components. The only such component I can think of, as @codemonkey800 pointed out, is the calendar, and it's not clear to me that this alone is sufficient motivation (though I do love our calendar).

Versioning does remain a big question mark in my opinion. What are our current plans for solving this issue? @kne42 mentions that sphinx has no multi-versioning capabilities. How are other projects doing this? How do we plan on doing it?

> That's what I meant. We could set up the theme into a pip package that you can just pip install and specify in the configuration. You don't need to worry about node or anything, you can just run jupyter-book build. The HTML will then be accessible to you fromfile:// (which is possible with the current theme if you do yarn build).

This is the holy grail imo. It should be of no relevance to someone contributing documentation what tooling the theme needs to be built, or how it is built. I think the theme should be built in whatever way is easiest for the developers, I just don't think that should change how we contribute documentation. *If* the theme can be easily installable for building and preview purposes, then the behind the scenes doesn't matter much to contributors.

> Personally, I don't feel like the usage documentation needs to match the hub theme (and I do kind of like matching the general look & feel of other scientific python packages)

I do think this is important. As the napari hub grows, it's quite possible people will find out about napari through the hub, or through one of its plugins. From that perspective I think cohesion is important. That being said, I *also* think it's important that pages like our API reference are familiar to people used to browsing other Python package docs.

**Comment by tlambert03** [comment]
> @kne42 mentions that sphinx has no multi-versioning capabilities. How are other projects doing this? How do we plan on doing it?

this isn't quite true, see for example how [pydata-sphinx-theme does it](https://pydata-sphinx-theme.readthedocs.io/en/latest/user_guide/configuring.html#add-a-dropdown-to-switch-between-docs-versions).  yes, it requires a bit of javascript, but it's pretty minimal.  [here](https://github.com/pydata/pydata-sphinx-theme/blob/master/src/pydata_sphinx_theme/theme/pydata_sphinx_theme/_templates/version-switcher.html) is their implementation.

**Comment by tlambert03** [comment]
I'll also point out that, readthedocs (which has it's own versioning feature) now support jupyter-book.  but, we'd need to have a pip-installable them there

**Comment by DragaDoncila** [comment]
> this isn't quite true, see for example how pydata-sphinx-theme does it. yes, it requires a bit of javascript, but it's pretty minimal. here is their implementation.

This seems pretty easy to implement 🤷‍♀️

**Comment by codemonkey800** [comment]
> The only such component I can think of, as @codemonkey800 pointed out, is the calendar, and it's not clear to me that this alone is sufficient motivation (though I do love our calendar).

I did write up a [doc](https://docs.google.com/document/d/10nkhIIlai9IHHtcepkMnG94IBwckHiaT1DgcL6wEg6Q/edit) sometime ago exploring React with napari.org. The features I mentioned in the doc were:

> - Table of Contents with automatic highlighting and collapsible sections.
> - Carousel of user submitted gifs demoing napari.
> - Installation matrix for different environments.
> - Unified search with the napari hub.

I'm not sure what other interactive features we'd want to add to naprai.org in the future, but given the list above, I felt it was appropriate to use React. It's hard to say because it's an investment on the future, but if we end up not implementing a lot of interactive features, then it's definitely overkill to use React.  @neuromusic @sofroniewn did y'all have other things in mind?

**Comment by codemonkey800** [comment]
The version switcher should be easy to implement regardless of whether we do it with DOM APIs or React. A React equivalent would look something like this in fewer LOC:

<details>
<summary>VersionDropdown.tsx</summary>

```tsx
import { useRouter } from 'next/router'
import { FormControl, InputLabel, MenuItem, Select } from '@material-ui/core'
import { useEffect, useState } from 'react'

import { useJupyterBookData } from '@/context/jupyterBook'

export function VersionDropdown() {
  const { currentVersion } = useJupyterBookData()
  const router = useRouter()
  const [versions, setVersions] = useState([])

  useEffect(() => {
    async function fetchVersions() {
      const response = await fetch('/versions.json')
      const data = await response.json()
      setVersions(data)
    }

    fetchVersions()
  }, [])

  async function versionRedirect() {
    const version = event.target.value

    const url = new URL(window.location.origin)
    url.pathname = `/${version}${router.pathname}`

    try {
      await fetch(url.href)
      window.location.href = url.href
    } catch (_) {
      url.pathname = `/${version}`
      window.location.href = url.href
    }
  }

  return (
    <FormControl>
      <InputLabel id="version-switcher" />
      <Select
        labelId="version-switcher"
        value={currentVersion}
        label="Version"
        onChange={versionRedirect}
      >
        {versions.map(version => (
          <MenuItem key={version} value={version}>
            {version}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  )
}

```

</details>

One docs site I like mentioning is [Botorch](https://botorch.org/) which uses [Docusuaurs](https://github.com/facebook/docusaurus) for managing docs and [built-in versionoing](https://docusaurus.io/docs/versioning), and Sphinx for generating the [API docs](https://botorch.org/api/). This might be complicated for our needs but it's a nice example of how we can blend the python and web ecosystems together, and that it doesn't necessarily have to follow what's standard for sphinx.

**Comment by tlambert03** [comment]
just to bring this back to the original point here, this isn't a referendum on react _per se_.  We can certainly debate whether react is "necessary" here, but this is ultimately about separation of concerns.  If you'd like to bundle react or other js libs in with your theme, that's fine.  But we shouldn't _depend_ on the theme for structure or content of the docs.  The theme shouldn't change the TOC.  There should be no problem with swapping out the theme for another theme (save, perhaps, a few special theme-specific features like a calendar, but those too should come with a theme directive or something that gets used in the markdown file).

**Comment by alisterburt** [comment]
This is some great discussion! Thanks for putting together concrete examples of how these things could look @tlambert03. One thing I would like to add which could potentially reframe the discussion a little bit...

napari.org is currently a mix between our docs and general project information, is separating these two things more explicitly something that has been considered? Most of the fancy features (calendar, download matrix) seem limited to the 'project site' side of things - this could be the 'hybrid approach' @codemonkey800 alluded to above.

Both [numpy](https://numpy.org/) and [pandas](https://pandas.pydata.org/) appear to have a larger separation between their 'project' site and their 'docs' site. Maybe separating concerns in this way could alleviate some of the issues around building docs and allow for the 'sandboxing' of dev tooling discussed above. I'm not sure how the numpy site is built, but their theming is more consistent between the numpy site and their docs than pandas, we could likely achieve a similar result with a docs specific sphinx theme for napari which matches the current react theme for the project site.

I'm also heavily in favour of 'docs should live in the repo if at all possible' but recognise the challenges - it seems like we can tackle this as a separate issue from the build pipeline though once a path forward has been agreed on?

Basically, what if we build docs completely separately from napari.org itself, then that static site gets pushed up to napari.org/docs which can have all of the fancy stuff?

**Comment by melissawm** [comment]
Just for clarity, NumPy uses three repos:
- API docs, how-tos, explanations are all in github.com/numpy/numpy, together with the code. These are pure sphinx, built with the pydata-sphinx-theme
- Tutorials, which are Jupyter Notebooks, live in github.com/numpy/numpy-tutorials. These are built using jupyterbook/myst-nb and jupytext.
- The numpy.org website, with landing page and a few informational pages, live in github.com/numpy/numpy.org

Those are all built independently, so contributions to one repo do not interfere with the others.

**Comment by neuromusic** [comment]
> [@tlambert03 ] just to bring this back to the original point here, this isn't a referendum on react per see.

+1 @tlambert03 ... from my understanding, the current structure was implemented in early 2021 and justified by the points that @kne42 highlights above. this was before we started using React.

> [@alisterburt ] napari.org is currently a mix between our docs and general project information, is separating these two things more explicitly something that has been considered? 

I think that this is definitely worth considering. Personally, I like the way that numpy handles the separation of concerns here.

> [@tlambert03] There should be no problem with swapping out the theme for another theme

I'm not sure that I agree with this, but if it is going to be a requirement it narrows the solution space immensely and is probably the most important decision to be made here.

**Comment by neuromusic** [comment]
one benefit of shipping the theme as a pip-installable sphinx theme: plugin devs could us it on their own documentation sites for a consistent look-and-feel with the napari ecosystem

**Comment by tlambert03** [comment]
Yep, I agree, it does narrow the solution space, at least for now.  But in my opinion, the broad solution space has not worked well for us, for quite a long time now; we've been experimenting with this pattern for over a year, there's been plenty of time to iterate, and it's still causing routine difficulties.  I'd like to see us form a solid working system before we start thinking about big plans.  It _may_ mean a small step backwards in scope for a little while, but that's what I'm proposing.  After we get a solid system, we can discuss how dependent on the theme we want to be.  As I mentioned above after the bit you quoted

> save, perhaps, a few special theme-specific features like a calendar, but those too should come with a theme directive or something that gets used in the markdown file

That is: all themes come with custom directives and stuff like that, so it's not like i see all themes as completely interchangeable.  But those bits should be clearly delineated, and – most importantly – the theme should not be dynamically organizing the content after build-time.

**Comment by codemonkey800** [comment]
> The theme shouldn't change the TOC. 

Furo does this too, just in Python:

- https://github.com/pradyunsg/furo/search?q=furo_navigation_Tree
- https://github.com/pradyunsg/furo/blob/d6b217f8bb021f137928aa30f1b41e829e87d73f/src/furo/navigation.py#L19-L87

> There should be no problem with swapping out the theme for another theme (save, perhaps, a few special theme-specific features like a calendar, but those too should come with a theme directive or something that gets used in the markdown file).

This is already possible: https://github.com/napari/napari.github.io/blob/main/_config.yml#L45-L47
If you remove `html_theme_path` and change `html_theme` to a different theme, the docs will build with that theme like any other sphinx theme. If we moved forward with moving to a pip installable package, then there wouldn't be a `html_theme_path`

**Comment by codemonkey800** [comment]
> But we shouldn't depend on the theme for structure or content of the docs.

This is a good point but I'm also not sure where to draw the line in the sand for. Up to now I felt the affordances provided by the theme have been fine, but perhaps we need to find what things should be provided by the content vs the theme

**Comment by kne42** [comment]
If we want to go through with this, I think it would be useful to think through whether we want only some parts of the site to have multiple versions or for the whole site to be versioned. Currently, in Lia's designs, some content is versioned (e.g. API docs), while other content is not (e.g. developer guides).

I'm bringing this to attention because there are some caveats if we were to combine versioned and unversioned content. For example, it would require that the versioned and unversioned content be built in Sphinx separately and then given the appearance of a combined look by having the same theme. PyTorch does this, for example, and they achieve it by in fact [hard-coding the top-level navigation into the sidebar](https://github.com/pytorch/pytorch_sphinx_theme/blob/b4d00058a48604d8fb63771b513a50450f0ee188/pytorch_sphinx_theme/layout.html#L105-L125).

Another issue that arises is that as the theme gets updated, previous builds of versioned docs can look out of place since they use the old theme. This happens if the full site is versioned too, but is more obvious when navigating a site where clicking on a link might bring you to a completely differently styled page.

I asked Eric Holscher, co-founder of Read The Docs, on how they addressed this and the following was his response:

> There's a few ways to handle it, but generally you need to rebuild all the versions if your theme has changed in some way. This can be done in a couple ways: the theme can be specified in a requirements file without a pinned version, the theme can be a submodule, or you can have some logic in your `conf.py` to update the theme to a defined version.

Overall, if we want to keep things simple, I would recommend versioning the full site because otherwise you run into other conundrums such as how to handle searching between versions from an unversioned page, linking to a versioned page from an unversioned page, etc.

**Comment by tlambert03** [comment]
I'm all for a full versioning... it's simpler, and most of the dev guides are being co-written along with a developing code base.  There'd be no need, for example, to show the current plugin or event reference docs (which are in the dev guides, not the API docs) with < v0.4.

**Comment by kne42** [comment]
Maybe dev guides was a bad example – I was referring more to things such as Mission & Values, How to Contribute, etc.

**Comment by tlambert03** [comment]
i feel like versioning those is relatively harmless, compared to the gain in simplicity.  It's even kind of nice to be able to see how these things evolve over time.  The vast majority of people will see the current release, and as long as we have the standard "you're looking at old docs" notice for those who aren't, it seems ok to me?

**Comment by sofroniewn** [comment]
The docs working group met to today to discuss this proposal. In attendance was @jni @tlambert03 @codemonkey800 @DragaDoncila @chili-chiu @nclack @dgmccart and @tdimino

To summarize:

1) A desirable goal is a pip installable package that comes pre-bundled with all of the javascript libraries it needs and doesn't require a local node environment to render the site.

2) Another goal is to improve the overall look and feel of the typography on the website - see https://github.com/napari/napari.github.io/issues/320 for specific details. We could either make changes to our current theme or match the look and feel we want by extending a sphinx theme like furo or sphinx-pydata.

The estimate for making these changes to the current theme (both pip install + typography) was ~4 to 6 weeks (though that is just an initial guess).

It has also been proposed that we could switch right away to using pydata or furo and begin modifying them. We could spend ~1 week making css changes so the color scheme matches what we have now and the transition is less of a jolt to users. We could then decide if we wanted to continue extending this theme or return to our more custom theme subject to the goals above.

@neuromusic @melissawm @GenevieveBuckley @liaprins-czi were unable to attend today, but we want to make sure to get your input here too.

cc @napari/core-devs as this is an important decision

**Comment by codemonkey800** [comment]
@jni @tlambert03 @sofroniewn should the theme live in it's own repo `napari/napari-sphinx-theme` or be in the same repo as the docs? having it in the same repo would make some aspects of development simpler, but we can work around it otherwise if separate repos is preferred

**Comment by melissawm** [comment]
I think the pip-installable theme is the best way to go :+1: There is still the question of image hosting and docs source location. I added a comment to explain what adopting git-lfs would mean for napari here: https://github.com/napari/napari/issues/3529#issuecomment-1017904329

**Comment by neuromusic** [comment]
As a path forward, discussed with @sofroniewn @justinelarsen & @jni and we propose the following...

1. Move forward to implement a new sphinx theme that will live in napari/napari-sphinx-theme that will implement the aspects of the current designs for napari.org that are meeting the needs of the community. We suspect the best path forward will be to start with a hard fork of furo or pydata as necessary for reasonable defaults where we haven't explicitly implemented custom designs (e.g. API docs).
2. While (1) is in development, to unblock community contribution to the docs, we'll rollback the designs to furo (with modifications of the color, etc as necessary) and work toward reorganizing the documentation infrastructure to be consistent with norms for python documentation (e.g. `make docs`, pip-installable theme, etc)

**Comment by sofroniewn** [comment]
There was further discussion of this topic at the community meeting today with @melissawm @DragaDoncila @liaprins-czi @codemonkey800 present too and weighing in.

It was agreed that there would be high-value in @codemonkey800 exploring a hard-fork of furo and how to bring that inline with the overall look and feel of the designs of @liaprins-czi while preserving the defaults for the API reference etc that we didn't do custom designs for. This work can be done in the newly created https://github.com/napari/napari-sphinx-theme repo. @liaprins-czi can work with @codemonkey800 to determine how best to make her design implementable with furo.

One advantage of this approach that @jni has called out is that it will allow us to get to the level of customization that we want gradually, which will make it easier to review changes, and easier for other folks in the community who want to customize their furo themes to see how we did made certain changes if they want to do them too.

If we realize that there are essential features of the design we won't be able to achieve building off furo then we can decide if we want to adopt what we are able to do with furo only as a stop-gap measure and build a custom pip installable sphinx theme ourselves

**Comment by neuromusic** [comment]
I've created an issue to track the development of the new sphinx theme: https://github.com/napari/napari-sphinx-theme/issues/1

This includes a link to a [Product Spec](https://hackmd.io/R6looyPgR7mNrHmPnPMtLQ?view) for the new sphinx theme. The Tech Spec is in progress & we'll review both at tomorrow's Documentation Working Group meeting, but feel free to review and weigh in on the docs or issue!

**Comment by tlambert03** [comment]
Thanks @neuromusic and everyone ... feels like we could probably close this issue now?  we've got enough other conversations and goal posts going on at this point.  If anyone wants to reopen that's fine.

**Comment by neuromusic** [comment]
Thanks for kicking off this convo @tlambert03!
